docs/rfc/active/graph_ql_frontend/mutations.md

```diff
@@ -0,0 +1,647 @@
+      - Generate `Mutation` types, omitting specific fields 
+      - Write generated types to GraphQL schema file 
+      - Run as a build-time gradle task
+
+We omit proposal of this portion of the design from this RFC. There will be a subsequent RFC proposing implementation of the items above.
+
+*Auto-generated GQL resolvers generation should also be explored. This would be possible provided that standardized DAOs are available at the frontend service layer. Specifically, it'd be incredible if we could
+- Auto generate "client" objects from a rest spec
+- Auto generate "dao" objects that use a client
+- Auto generate "resolvers" that use a "dao" 
+
+### Executing a Mutation
+
+For executing mutations, we will rely on the [graphql-java](https://www.graphql-java.com/) library.
+
+As a prerequisite, we'll need to 
+
+1. implement a GraphQL resolver 
+2. register the GraphQL resolver
+
+for each entity + operation combination. The concept of resolvers has been already introduced in RFC 2042. 
+
+**The Approach** 
+
+In the case of mutations, resolvers will be responsible for performing an atomic update of an entity's state. This may involve simultaneously updating different aspects for the same entity. To achieve this without the need for supporting distributed transactions, we'll leverage the "ingest" API exposed by GMA for each top-level entity. This API accepts an entity `snapshot` (eg. `DatasetSnapshot`) and updates each aspect included within. 
+
+This approach stands in contrast to the existing recommendation around updating an entity's aspects via the GMA Rest.li API. The current recommendation involves writing to aspect-oriented Rest.li sub-resources (eg. datasets/:urn/ownership). This approach means that each request can only change one aspect. Because the proposed GQL API permits mutation of multiple entity aspects within a mutation query, adopting this approach would require implementing a transaction layer capable of preventing inconsistent states. Moreover, it would require more extensive changes across the stack to introduce new aspects to existing entities.
+
+*Folks from DH have noted that the entity "ingest" API is intended to be private, only used by the MCE consumer logic. We propose formal support for this API (or something of equivalent function) for atomically updating multiple entity aspects at one time.
+
+*A related point worth considering is the modeling inconsistency in APIs exposed by GMA. Today, we are able to *read* entity models like `Dataset.pdl` but are unable to *write* using this model. We should aim for clarity & consistency in the model exposed to clients of GMA by exposing `Entity` or `Snapshot` models for both read *&* write -- not a mixture of both. To work around this inconsistency, the sample code provided with this RFC includes entity-to-snapshot conversion hidden within the `Datasets` DAO, thereby allowing the API layer to concern itself only with the canonical entity model.  
```

Comment:
In either case I think this API should be considerate of how it would look if was the replacement for restli, so that we can easily replace restli with it in the future. The resolvers, rather than read/write via the restli APIs, would just read/write to DBs. That migration be transparent to anyone using GQL before the migration, if we can keep the APIs the same.

Reply:
Maybe we should have both aspects and entities be readable / writable?

The entity code would use GQL to read/write the aspect GQL, which turn turn would read/write to the restli endpoints (though this also requires every aspect have a restli endpoint, which I don't think is true today?), or in the future right to the DB. So we could have the entity layer "on top of" the aspect layer. While this doesn't get you distributed extensions, it does more closely reflect the underlying data model, and what we have today.

Not sure if I like it or not, just a thought.